safe-api/src/api/app/xorurl.rs

```
@@ -289,92 +536,350 @@ impl XorUrlEncoder {
+    ///    blue would be returned when key is "color".
+    pub fn query_key_last(&self, key: &str) -> Option<String> {
+        Self::query_key_last_internal(&self.query_string, key)
+    }
+
+    /// returns the first matching key from a query string.
+    ///
+    /// eg in safe://name?color=red&age=5&color=green&color=blue
+    ///    red would be returned when key is "color".
+    pub fn query_key_first(&self, key: &str) -> Option<String> {
+        Self::query_key_first_internal(&self.query_string, key)
     }
 
+    /// sets url fragment
     pub fn set_fragment(&mut self, fragment: String) {
         self.fragment = fragment;
     }
 
+    /// Retrieves url fragment, without # separator
+    pub fn fragment(&self) -> String {
+        self.fragment.clone()
+    }
+
+    /// Retrieves url fragment, with # separator if non-empty.
+    pub fn fragment_with_separator(&self) -> String {
+        if self.fragment.is_empty() {
+            "".to_string()
+        } else {
+            format!("#{}", self.fragment)
+        }
+    }
+
+    /// returns true if an NrsUrl, false if an XorUrl
+    pub fn is_nrs(&self) -> bool {
+        !self.nrs_host.is_empty()
+    }
+
     // XOR-URL encoding format (var length from 36 to 44 bytes):
     // 1 byte for encoding version
     // 2 bytes for content type (enough to start including some MIME types also)
     // 1 byte for SAFE native data type
     // 32 bytes for XoR Name
     // and up to 8 bytes for type_tag
     // query param "v=" is treated as the content version
-    pub fn to_string(&self) -> Result<String> {
-        self.to_base(DEFAULT_XORURL_BASE)
+
+    /// serializes the URL to an XorUrl string.
+    ///
+    /// This function may be called on an NrsUrl and
+    /// the corresponding XorUrl will be returned.
+    pub fn to_xorurl_string(&self) -> String {
+        self.to_base(DEFAULT_XORURL_BASE).unwrap_or_else(|e| {
+            warn!("{}", e);
```

I don't think this is a desirable way of not returning an error, we are preventing the consumer from knowing what happened, when I call this function, since it doesn't return a Result<>, I then assume it will never fail but it can fail and it doesn't even tell me so, using an empty string as the way to check errors is going against the language way of handling errors properly.

---

So I just reviewed again and the *only* place that can generate an error within this func is inside host_to_base() where it does a check that MediaType is not unknown.  This same/similar check is performed in ::new(), and there is no content_type setter.  In other words, if an error occurs, there is a bug in the implementation.  I reasoned that in this situation, it is acceptable to log warning and return an empty string.   The alternatives would seem to be:
 *  returning the error to caller, which we know/believe should never actually hit the error case.  
 * getting rid of the check in name_to_base(), since it is redundant with check in ::new().

I think I would prefer the latter, adding a comment that the validation is performed in new()

what do you think?

in ::host_to_base()
```
            SafeContentType::MediaType(media_type) => match MEDIA_TYPE_CODES.get(media_type) {
                Some(media_type_code) => *media_type_code,
                None => {
                    return Err(Error::Unexpected(format!(
                        "Failed to encode Media-type '{}'",
                        media_type
                    )))
                }
            },
```

in ::new()
```
        if let SafeContentType::MediaType(ref media_type) = content_type {
            if !Self::is_media_type_supported(media_type) {
                return Err(Error::InvalidMediaType(format!(
                        "Media-type '{}' not supported. You can use 'SafeContentType::Raw' as the 'content_type' for this type of content",
                        media_type
                    )));
            }
        }

```

---

Yes, I agree, the latter is the best option and what is usually done, any setter or constructor should take care of validations and return corresponding errors, so we can then have all type of getters (like to_string() or to_base()) to return a value and not a Result<>. Note we also are using the word host in that function that we need to rename/remove.

> it is acceptable to log warning and return an empty string
I think this is never acceptable, you are not handling error accordingly and hiding it. If you know it will never cause an error should not return a Result from a function, and it should always return a valid value. In this case you are forced to unwrap due to host_to_base() returning a Result<> which means it could fail, so you cannot ignore an error from that result/call, anyone making changes to the host_to_base() function may return a valid error as he/she sees the function returns a Result which means it's expected to return errors, but we won't be handling them correctly. Hiding "unexpected" errors like that is giving up the compilation check (which is why Result exists I believe), so you'll be having a bug sooner or later due to it.

---

Looking at this MediaType code more closely, I see that the problem is that MediaType accepts an arbitrary String, eg "text/html", but then host_to_base() tries to look it up in a list of supported media types, in order to convert to a u16.  This lookup can return None, so there is no way to not handle the lookup failure.  I can think of a few possibilities:

1. We perform the lookup in ::new(), which is being done anyway, and save the u16 in the SafeUrl struct, else throw error.   only downside is that this requires extra space in the struct.
2. We ignore the possibility of error, because we know it is validated elsewhere. 
3. ::to_xorurl_string() must return a Result.

So there are tradeoffs with every possibility.  I tend to favor #1.  An extra u16 in the struct doesn't seem that bad.

---

it sounds good to me, you will be having a u16 but you will be getting rid of the string you are storing now in the internal content_type field.

---

ok, I implemented #1.  works fine.

> it sounds good to me, you will be having a u16 but you will be getting rid of the string you are storing now in the internal content_type field.

Unfortunately I could not remove the existing content_type field because it is actually a SafeContentType and is returned as such from ::content_type().  One might think it could be constructed from the u16, but in practice, SafeContentType::from_u16() can throw an error if input u16 is invalid.  (aside: looking at the ::from_value() routine, it doesn't actually support MediaType at all, so is incomplete.)  So the only way to keep content_type() API as it is, and remove result from to_xorurl_string() without ignoring any errors is to store both the SafeContentType and correesponding/validated u16 in the struct, which is a little gross, but works.